README.md
LightFace
=========

LightFace is a clone of FaceBook's modal dialog.  LightFace offers numerous options to customize each modal instance.  LightFace also provides utility classes specially designed to work with AJAX requests, IFrames, and images.

![Screenshot](http://davidwalsh.name/dw-content/lightface.png)


How to Use
----------

LightFace instances can be created at any time.  There are no arguments other than the instance options.

	#JS
	/* create LightFace instance */
	var modal = new LightFace({
		width: "auto",
		height: 400,
		draggable: true,
		title: "Hello from LightFace!",
		content: "<p>This is the LightFace content!</p>",
		buttons: [
			{ title: "Close", event: function() { this.close(); }, color: "blue" }
		],
		resetOnScroll: true
	});

	/* open when link is clicked */
	document.id("launchModal").addEvent("click", function(e){
		e.stop();
		modal.open();
	});
	
LightFace works great with static content, IFrames, AJAX Requests, tooltips, and images.  LightFace also resizes and repositions when the window is resized or scrolled.  Tested in IE7+, FF, Safari, Opera, Chrome, and iPad.

For specific usage and options, please read the documentation or visit [http://davidwalsh.name/js/lightface](http://davidwalsh.name/js/lightface)

changelog.md
LightFace
=========

The following changes have been made to LightFace.

03/21/2012
----------
	+ Updated overlay to have visibility:hidden to work best with Moo 1.4+
	+ Tag updated from .96 to .97

11/17/2010
----------
	+ Updated destroy() method to remove reference to node so that videos stop playing when destroy() is called.
	+ Tag updated from .94 to .96

tests/ajax.php
<?php 
	sleep(1);
	echo 'Hello '.$_POST['name']; 
?>

tests/mootools-more-drag.js
// MooTools: the javascript framework.
// Load this file's selection again by visiting: http://mootools.net/more/d4e0cddc091a13e814dd53a6609b5cce 
// Or build this file again with packager using: packager build More/Drag
/*
---

script: More.js

name: More

description: MooTools More

license: MIT-style license

authors:
  - Guillermo Rauch
  - Thomas Aylott
  - Scott Kyle
  - Arian Stolwijk
  - Tim Wienk
  - Christoph Pojer
  - Aaron Newton

requires:
  - Core/MooTools

provides: [MooTools.More]

...
*/

MooTools.More = {
	'version': '1.3.0.1',
	'build': '6dce99bed2792dffcbbbb4ddc15a1fb9a41994b5'
};


/*
---

script: Drag.js

name: Drag

description: The base Drag Class. Can be used to drag and resize Elements using mouse events.

license: MIT-style license

authors:
  - Valerio Proietti
  - Tom Occhinno
  - Jan Kassens

requires:
  - Core/Events
  - Core/Options
  - Core/Element.Event
  - Core/Element.Style
  - Core/Element.Dimensions
  - /MooTools.More

provides: [Drag]
...

*/

var Drag = new Class({

	Implements: [Events, Options],

	options: {/*
		onBeforeStart: function(thisElement){},
		onStart: function(thisElement, event){},
		onSnap: function(thisElement){},
		onDrag: function(thisElement, event){},
		onCancel: function(thisElement){},
		onComplete: function(thisElement, event){},*/
		snap: 6,
		unit: 'px',
		grid: false,
		style: true,
		limit: false,
		handle: false,
		invert: false,
		preventDefault: false,
		stopPropagation: false,
		modifiers: {x: 'left', y: 'top'}
	},

	initialize: function(){
		var params = Array.link(arguments, {
			'options': Type.isObject,
			'element': function(obj){
				return obj != null;
			}
		});

		this.element = document.id(params.element);
		this.document = this.element.getDocument();
		this.setOptions(params.options || {});
		var htype = typeOf(this.options.handle);
		this.handles = ((htype == 'array' || htype == 'collection') ? $$(this.options.handle) : document.id(this.options.handle)) || this.element;
		this.mouse = {'now': {}, 'pos': {}};
		this.value = {'start': {}, 'now': {}};

		this.selection = (Browser.ie) ? 'selectstart' : 'mousedown';


		if (Browser.ie && !Drag.ondragstartFixed){
			document.ondragstart = Function.from(false);
			Drag.ondragstartFixed = true;
		}

		this.bound = {
			start: this.start.bind(this),
			check: this.check.bind(this),
			drag: this.drag.bind(this),
			stop: this.stop.bind(this),
			cancel: this.cancel.bind(this),
			eventStop: Function.from(false)
		};
		this.attach();
	},

	attach: function(){
		this.handles.addEvent('mousedown', this.bound.start);
		return this;
	},

	detach: function(){
		this.handles.removeEvent('mousedown', this.bound.start);
		return this;
	},

	start: function(event){
		var options = this.options;

		if (event.rightClick) return;

		if (options.preventDefault) event.preventDefault();
		if (options.stopPropagation) event.stopPropagation();
		this.mouse.start = event.page;

		this.fireEvent('beforeStart', this.element);

		var limit = options.limit;
		this.limit = {x: [], y: []};

		var styles = this.element.getStyles('left', 'right', 'top', 'bottom');
		this._invert = {
			x: options.modifiers.x == 'left' && styles.left == 'auto' && !isNaN(styles.right.toInt()) && (options.modifiers.x = 'right'),
			y: options.modifiers.y == 'top' && styles.top == 'auto' && !isNaN(styles.bottom.toInt()) && (options.modifiers.y = 'bottom')
		};

		var z, coordinates;
		for (z in options.modifiers){
			if (!options.modifiers[z]) continue;

			var style = this.element.getStyle(options.modifiers[z]);

			// Some browsers (IE and Opera) don't always return pixels.
			if (style && !style.match(/px$/)){
				if (!coordinates) coordinates = this.element.getCoordinates(this.element.getOffsetParent());
				style = coordinates[options.modifiers[z]];
			}

			if (options.style) this.value.now[z] = (style || 0).toInt();
			else this.value.now[z] = this.element[options.modifiers[z]];

			if (options.invert) this.value.now[z] *= -1;
			if (this._invert[z]) this.value.now[z] *= -1;

			this.mouse.pos[z] = event.page[z] - this.value.now[z];

			if (limit && limit[z]){
				var i = 2;
				while (i--){
					var limitZI = limit[z][i];
					if (limitZI || limitZI === 0) this.limit[z][i] = (typeof limitZI == 'function') ? limitZI() : limitZI;
				}
			}
		}

		if (typeOf(this.options.grid) == 'number') this.options.grid = {
			x: this.options.grid,
			y: this.options.grid
		};

		var events = {
			mousemove: this.bound.check,
			mouseup: this.bound.cancel
		};
		events[this.selection] = this.bound.eventStop;
		this.document.addEvents(events);
	},

	check: function(event){
		if (this.options.preventDefault) event.preventDefault();
		var distance = Math.round(Math.sqrt(Math.pow(event.page.x - this.mouse.start.x, 2) + Math.pow(event.page.y - this.mouse.start.y, 2)));
		if (distance > this.options.snap){
			this.cancel();
			this.document.addEvents({
				mousemove: this.bound.drag,
				mouseup: this.bound.stop
			});
			this.fireEvent('start', [this.element, event]).fireEvent('snap', this.element);
		}
	},

	drag: function(event){
		var options = this.options;

		if (options.preventDefault) event.preventDefault();
		this.mouse.now = event.page;

		for (var z in options.modifiers){
			if (!options.modifiers[z]) continue;
			this.value.now[z] = this.mouse.now[z] - this.mouse.pos[z];

			if (options.invert) this.value.now[z] *= -1;
			if (this._invert[z]) this.value.now[z] *= -1;

			if (options.limit && this.limit[z]){
				if ((this.limit[z][1] || this.limit[z][1] === 0) && (this.value.now[z] > this.limit[z][1])){
					this.value.now[z] = this.limit[z][1];
				} else if ((this.limit[z][0] || this.limit[z][0] === 0) && (this.value.now[z] < this.limit[z][0])){
					this.value.now[z] = this.limit[z][0];
				}
			}

			if (options.grid[z]) this.value.now[z] -= ((this.value.now[z] - (this.limit[z][0]||0)) % options.grid[z]);

			if (options.style) this.element.setStyle(options.modifiers[z], this.value.now[z] + options.unit);
			else this.element[options.modifiers[z]] = this.value.now[z];
		}

		this.fireEvent('drag', [this.element, event]);
	},

	cancel: function(event){
		this.document.removeEvents({
			mousemove: this.bound.check,
			mouseup: this.bound.cancel
		});
		if (event){
			this.document.removeEvent(this.selection, this.bound.eventStop);
			this.fireEvent('cancel', this.element);
		}
	},

	stop: function(event){
		var events = {
			mousemove: this.bound.drag,
			mouseup: this.bound.stop
		};
		events[this.selection] = this.bound.eventStop;
		this.document.removeEvents(events);
		if (event) this.fireEvent('complete', [this.element, event]);
	}

});

Element.implement({

	makeResizable: function(options){
		var drag = new Drag(this, Object.merge({
			modifiers: {
				x: 'width',
				y: 'height'
			}
		}, options));

		this.store('resizer', drag);
		return drag.addEvent('drag', function(){
			this.fireEvent('resize', drag);
		}.bind(this));
	}

});



Source/LightFace.Image.js
/*
---
description:     LightFace.Image

authors:
  - David Walsh (http://davidwalsh.name)

license:
  - MIT-style license

requires:
  - LightFace

provides:
  - LightFace.Image
...
*/
LightFace.Image = new Class({
	Extends: LightFace,
	options: {
		constrain: true,
		url: ''
	},
	initialize: function(options) {
		this.parent(options);
		this.url = '';
		this.resizeOnOpen = false;
		if(this.options.url) this.load();
	},
	_resize: function() {
		//get the largest possible height
		var maxHeight = window.getSize().y - this.options.pad;

		//get the image size
		var imageDimensions = document.id(this.image).retrieve('dimensions');

		//if image is taller than window...
		if(imageDimensions.y > maxHeight) {
			this.image.height = maxHeight;
			this.image.width = (imageDimensions.x * (maxHeight / imageDimensions.y));
			this.image.setStyles({
				height: maxHeight,
				width: (imageDimensions.x * (maxHeight / imageDimensions.y)).toInt()
			});
		}

		//get rid of styles
		this.messageBox.setStyles({ height: '', width: '' });

		//position the box
		this._position();
	},
	load: function(url,title) {
		//keep current height/width
		var currentDimensions = { x: '', y: '' };
		if(this.image) currentDimensions = this.image.getSize();
		///empty the content, show the indicator
		this.messageBox.set('html','').addClass('lightFaceMessageBoxImage').setStyles({
			width: currentDimensions.x,
			height: currentDimensions.y
		});
		this._position();
		this.fade();
		this.image = new Element('img',{
			events: {
				load: function() {
					(function() {
						var setSize = function() { this.image.inject(this.messageBox).store('dimensions',this.image.getSize()); }.bind(this);
						setSize();
						this._resize();
						setSize(); //stupid ie
						this.unfade();
						this.fireEvent('complete');
					}).bind(this).delay(10);
				}.bind(this),
				error: function() {
					this.fireEvent('error');
					this.image.destroy();
					delete this.image;
					this.messageBox.set('html',this.options.errorMessage).removeClass('lightFaceMessageBoxImage');
				}.bind(this),
				click: function() {
					this.close();
				}.bind(this)
			},
			styles: {
				width: 'auto',
				height: 'auto'
			}
		});
		this.image.src = url || this.options.url;
		if(title && this.title) this.title.set('html',title);
		return this;
	}
});

Source/LightFace.IFrame.js
/*
---
description:     LightFace.IFrame

authors:
  - David Walsh (http://davidwalsh.name)

license:
  - MIT-style license

requires:
  - LightFace

provides:
  - LightFace.IFrame
...
*/
LightFace.IFrame = new Class({
	Extends: LightFace,
	options: {
		url: ''
	},
	initialize: function(options) {
		this.parent(options);
		if(this.options.url) this.load();
	},
	load: function(url,title) {
		this.fade();
		if(!this.iframe) {
			this.messageBox.set('html','');
			this.iframe = new IFrame({
				styles: {
					width: '100%',
					height: '100%'
				},
				events: {
					load: function() {
						this.unfade();
						this.fireEvent('complete');
					}.bind(this)
				},
				border: 0
			}).inject(this.messageBox);
			this.messageBox.setStyles({ padding:0, overflow:'hidden' });
		}
		if(title) this.title.set('html',title);
		this.iframe.src = url || this.options.url;
		this.fireEvent('request');
		return this;
	}
});

Source/LightFace.Request.js
/*
---
description:     LightFace.Request

authors:
  - David Walsh (http://davidwalsh.name)

license:
  - MIT-style license

requires:
  - LightFace

provides:
  - LightFace.Request
...
*/
LightFace.Request = new Class({
	Extends: LightFace,
	options: {
		url: '',
		request: {
			url: false
		}
	},
	initialize: function(options) {
		this.parent(options);
		if(this.options.url) this.load();
	},
	load: function(url,title) {
		var props = (Object.append || $extend)({
			onRequest: function() {
				this.fade();
				this.fireEvent('request');
			}.bind(this),
			onSuccess: function(response) {
				this.messageBox.set('html',response);
				this.fireEvent('success');
			}.bind(this),
			onFailure: function() {
				this.messageBox.set('html',this.options.errorMessage);
				this.fireEvent('failure');
			}.bind(this),
			onComplete: function() {
				this._resize();
				this._ie6Size();
				this.messageBox.setStyle('opacity',1);
				this.unfade();
				this.fireEvent('complete');
			}.bind(this)
		},this.options.request);

		if(title && this.title) this.title.set('html',title);
		if(!props.url) props.url = url || this.options.url;

		new Request(props).send();
		return this;
	}
});

Source/LightFace-yui-compressed.js
/*
	Inspired by Bert Ramaker's Facebox for jQuery and MooTools:  http://bertramakers.com/labs/
	Rewritten and optimized by David Walsh:  http://davidwalsh.name
*/
/*
---
description:     LightFace

authors:
  - David Walsh (http://davidwalsh.name)

license:
  - MIT-style license

requires:
  core/1.2.1:   '*'

provides:
  - LightFace
  - LightFace.IFrame
  - LightFace.Image
  - LightFace.Request
  - LightFace.Static
...
*/
var LightFace=new Class({Implements:[Options,Events],options:{width:"auto",height:"auto",draggable:false,title:"",buttons:[],fadeDelay:400,fadeDuration:400,keys:{esc:function(){this.close()}},content:"<p>Message not specified.</p>",zIndex:9001,pad:100,overlayAll:false,constrain:false,resetOnScroll:true,baseClass:"lightface",errorMessage:"<p>The requested file could not be found.</p>"},initialize:function(a){this.setOptions(a);this.state=false;this.buttons={};this.resizeOnOpen=true;this.ie6=typeof document.body.style.maxHeight=="undefined";this.draw()},draw:function(){this.box=new Element("table",{"class":this.options.baseClass,styles:{"z-index":this.options.zIndex,opacity:0},tween:{duration:this.options.fadeDuration,onComplete:function(){if(this.box.getStyle("opacity")==0){this.box.setStyles({top:-9000,left:-9000})}}.bind(this)}}).inject(document.body,"bottom");var h=["top","center","bottom"],e=["Left","Center","Right"],c=h.length;for(var b=0;b<c;b++){var f=this.box.insertRow(b);for(var g=0;g<c;g++){var d=h[b]+e[g],a=f.insertCell(g);a.className=d;if(d=="centerCenter"){this.contentBox=new Element("div",{"class":"lightfaceContent",styles:{width:this.options.width}});a.appendChild(this.contentBox)}else{document.id(a).setStyle("opacity",0.4)}}}if(this.options.title){this.title=new Element("h2",{"class":"lightfaceTitle",html:this.options.title}).inject(this.contentBox);if(this.options.draggable&&$defined(window.Drag)){this.draggable=true;new Drag(this.box,{handle:this.title});this.title.addClass("lightfaceDraggable")}}this.messageBox=new Element("div",{"class":"lightfaceMessageBox",html:this.options.content||"",styles:{height:this.options.height}}).inject(this.contentBox);this.footer=new Element("div",{"class":"lightfaceFooter",styles:{display:"none"}}).inject(this.contentBox);this.overlay=new Element("div",{html:"&nbsp;",styles:{opacity:0},"class":"lightfaceOverlay",tween:{link:"chain",duration:this.options.fadeDuration,onComplete:function(){if(this.overlay.getStyle("opacity")==0){this.box.focus()}}.bind(this)}}).inject(this.contentBox);if(!this.options.overlayAll){this.overlay.setStyles({top:(this.title?this.title.getSize().y:0)})}this.buttons=[];if(this.options.buttons.length){this.options.buttons.each(function(i){this.addButton(i.title,i.event,i.color)},this)}this.focusNode=this.box;return this},addButton:function(e,a,b){this.footer.setStyle("display","block");var d="lightfacefocus"+b;var c=new Element("label",{"class":b?"lightface"+b:"",events:{mousedown:function(){if(b){c.addClass(d);var f=function(){c.removeClass(d);document.id(document.body).removeEvent("mouseup",f)};document.id(document.body).addEvent("mouseup",f)}}}});this.buttons[e]=(new Element("input",{type:"button",value:e,events:{click:(a||this.close).bind(this)}}).inject(c));c.inject(this.footer);return this},showButton:function(a){if(this.buttons[a]){this.buttons[a].removeClass("hiddenButton")}return this.buttons[a]},hideButton:function(a){if(this.buttons[a]){this.buttons[a].addClass("hiddenButton")}return this.buttons[a]},close:function(a){if(this.isOpen){this.box[a?"setStyles":"tween"]("opacity",0);this.fireEvent("close");this._detachEvents();this.isOpen=false}return this},open:function(a){if(!this.isOpen){this.box[a?"setStyles":"tween"]("opacity",1);if(this.resizeOnOpen){this._resize()}this.fireEvent("open");this._attachEvents();(function(){this._setFocus()}).bind(this).delay(this.options.fadeDuration+10);this.isOpen=true}return this},_setFocus:function(){this.focusNode.setAttribute("tabIndex",0);this.focusNode.focus()},fade:function(b,a){this._ie6Size();(function(){this.overlay.setStyle("opacity",b||1)}.bind(this)).delay(a||0);this.fireEvent("fade");return this},unfade:function(a){(function(){this.overlay.fade(0)}.bind(this)).delay(a||this.options.fadeDelay);this.fireEvent("unfade");return this},_ie6Size:function(){if(this.ie6){var a=this.contentBox.getSize();var b=(this.options.overlayAll||!this.title)?0:this.title.getSize().y;this.overlay.setStyles({height:a.y-b,width:a.x})}},load:function(a,b){if(a){this.messageBox.set("html",a)}if(b&&this.title){this.title.set("html",b)}this.fireEvent("complete");return this},_attachEvents:function(){this.keyEvent=function(a){if(this.options.keys[a.key]){this.options.keys[a.key].call(this)}}.bind(this);this.focusNode.addEvent("keyup",this.keyEvent);this.resizeEvent=this.options.constrain?function(a){this._resize()}.bind(this):function(){this._position()}.bind(this);window.addEvent("resize",this.resizeEvent);if(this.options.resetOnScroll){this.scrollEvent=function(){this._position()}.bind(this);window.addEvent("scroll",this.scrollEvent)}return this},_detachEvents:function(){this.focusNode.removeEvent("keyup",this.keyEvent);window.removeEvent("resize",this.resizeEvent);if(this.scrollEvent){window.removeEvent("scroll",this.scrollEvent)}return this},_position:function(){var a=window.getSize(),b=window.getScroll(),c=this.box.getSize();this.box.setStyles({left:b.x+((a.x-c.x)/2),top:b.y+((a.y-c.y)/2)});this._ie6Size();return this},_resize:function(){var b=this.options.height;if(b=="auto"){var a=window.getSize().y-this.options.pad;if(this.contentBox.getSize().y>a){b=a}}this.messageBox.setStyle("height",b);this._position()},toElement:function(){return this.messageBox},getBox:function(){return this.box},destroy:function(){this._detachEvents();this.button.each(function(a){a.removeEvents("click")});this.box.dispose()}});LightFace.IFrame=new Class({Extends:LightFace,options:{url:""},initialize:function(a){this.parent(a);if(this.options.url){this.load()}},load:function(a,b){this.fade();if(!this.iframe){this.messageBox.set("html","");this.iframe=new IFrame({styles:{width:"100%",height:"100%"},events:{load:function(){this.unfade();this.fireEvent("complete")}.bind(this)},border:0}).inject(this.messageBox);this.messageBox.setStyles({padding:0,overflow:"hidden"})}if(b){this.title.set("html",b)}this.iframe.src=a||this.options.url;this.fireEvent("request");return this}});LightFace.Image=new Class({Extends:LightFace,options:{constrain:true,url:""},initialize:function(a){this.parent(a);this.url="";this.resizeOnOpen=false;if(this.options.url){this.load()}},_resize:function(){var b=window.getSize().y-this.options.pad;var a=document.id(this.image).retrieve("dimensions");if(a.y>b){this.image.height=b;this.image.width=(a.x*(b/a.y));this.image.setStyles({height:b,width:(a.x*(b/a.y)).toInt()})}this.messageBox.setStyles({height:"",width:""});this._position()},load:function(a,c){var b={x:"",y:""};if(this.image){b=this.image.getSize()}this.messageBox.set("html","").addClass("lightFaceMessageBoxImage").setStyles({width:b.x,height:b.y});this._position();this.fade();this.image=new Element("img",{events:{load:function(){(function(){var d=function(){this.image.inject(this.messageBox).store("dimensions",this.image.getSize())}.bind(this);d();this._resize();d();this.unfade();this.fireEvent("complete")}).bind(this).delay(10)}.bind(this),error:function(){this.fireEvent("error");this.image.destroy();delete this.image;this.messageBox.set("html",this.options.errorMessage).removeClass("lightFaceMessageBoxImage")}},style:"width:auto;height:auto;"});this.image.src=a||this.options.url;if(c&&this.title){this.title.set("html",c)}return this}});LightFace.Request=new Class({Extends:LightFace,options:{url:"",request:{url:false}},initialize:function(a){this.parent(a);if(this.options.url){this.load()}},load:function(a,c){var b=$extend({onRequest:function(){this.fade();this.fireEvent("request")}.bind(this),onSuccess:function(d){this.messageBox.set("html",d);this.fireEvent("success")}.bind(this),onFailure:function(){this.messageBox.set("html",this.options.errorMessage);this.fireEvent("failure")}.bind(this),onComplete:function(){this._resize();this.messageBox.setStyle("opacity",1);this.unfade();this.fireEvent("complete")}.bind(this)},this.options.request);if(c&&this.title){this.title.set("html",c)}if(!b.url){b.url=a||this.options.url}new Request(b).send();return this}});LightFace.Static=new Class({Extends:LightFace,options:{offsets:{x:20,y:20}},open:function(b,a,c){this.parent(b);this._position(a,c)},_position:function(a,b){if(!$defined(a)){return}this.box.setStyles({top:b-this.options.offsets.y,left:a-this.options.offsets.x})}});

Source/LightFace.js
/*
---
description: LightFace

license: MIT-style

authors:
- David Walsh (http://davidwalsh.name)

requires:
- core/1.2.1: '*'

provides: [LightFace]

...
*/

var LightFace = new Class({
	
	Implements: [Options,Events],
	
	options: {
		width: 'auto',
		height: 'auto',
		destroyOnClose: false,
		draggable: false,
		title: '',
		buttons: [],
		fadeDelay: 400,
		fadeDuration: 400,
		keys: { 
			esc: function() { this.close(); } 
		},
		content: '<p>Message not specified.</p>',
		zIndex: 9001,
		pad: 100,
		overlayAll: false,
		constrain: false,
		resetOnScroll: true,
		baseClass: 'lightface',
		errorMessage: '<p>The requested file could not be found.</p>',
		/* Mask Options */
		closeOnMaskClick: false,
		maskClass: 'lightfaceMask',
		showMask: true/*,
		onOpen: $empty,
		onClose: $empty,
		onFade: $empty,
		onUnfade: $empty,
		onComplete: $empty,
		onRequest: $empty,
		onSuccess: $empty,
		onFailure: $empty
		*/
	},
	
	
	initialize: function(options) {
		this.setOptions(options);
		this.state = false;
		this.buttons = {};
		this.resizeOnOpen = true;
		this.ie6 = typeof document.body.style.maxHeight == "undefined";
		this.oMask = null;
		this.oParent = null; /* Parent LightFace object */
		this.dependents = []; /* Dependent LightFace objects */
		this.draw();
	},
	
	draw: function() {

		//create modal mask
		if (this.options.showMask && window['Mask'] != null) {
			this.oMask = new Mask(null, {
				'class': this.options.maskClass,
				onClick: function() {
					if (this.options.closeOnMaskClick) this.close();
				}.bind(this)
			});
		}
		
		//create main box
		this.box = new Element('table',{
			'class': this.options.baseClass,
			styles: {
				'z-index': this.options.zIndex,
				opacity: 0,
				visibility: 'hidden'
			},
			tween: {
				duration: this.options.fadeDuration,
				onComplete: function() {
					if(this.box.getStyle('opacity') == 0) {
						this.box.setStyles({ top: -9000, left: -9000 });
					}
				}.bind(this)
			}
		}).inject(document.body,'bottom');

		//draw rows and cells;  use native JS to avoid IE7 and I6 offsetWidth and offsetHeight issues
		var verts = ['top','center','bottom'], hors = ['Left','Center','Right'], len = verts.length;
		for(var x = 0; x < len; x++) {
			var row = this.box.insertRow(x);
			for(var y = 0; y < len; y++) {
				var cssClass = verts[x] + hors[y], cell = row.insertCell(y);
				cell.className = cssClass;
				if (cssClass == 'centerCenter') {
					this.contentBox = new Element('div',{
						'class': 'lightfaceContent',
						styles: {
							width: this.options.width
						}
					});
					cell.appendChild(this.contentBox);
				}
				else {
					document.id(cell).setStyles({ visibility: 'visible', opacity: 0.4 });
				}
			}
		}
		
		//draw title
		
		this.title = new Element('h2',{
			'class': 'lightfaceTitle',
			html: this.options.title
		}).inject(this.contentBox);

		if(this.options.draggable && window['Drag'] != null) {
			this.draggable = true;
			new Drag(this.box,{ handle: this.title });
			this.title.addClass('lightfaceDraggable');
		}
		
		//draw message box
		this.messageBox = new Element('div',{
			'class': 'lightfaceMessageBox',
			html: this.options.content || '',
			styles: {
				height: this.options.height
			}
		}).inject(this.contentBox);

		//button container
		this.footer = new Element('div',{
			'class': 'lightfaceFooter',
			styles: {
				display: 'none'
			}
		}).inject(this.contentBox);
		
		//draw overlay
		this.overlay = new Element('div',{
			html: '&nbsp;',
			styles: {
				opacity: 0,
				visibility: 'hidden'
			},
			'class': 'lightfaceOverlay',
			tween: {
				link: 'chain',
				duration: this.options.fadeDuration,
				onComplete: function() {
					if(this.overlay.getStyle('opacity') == 0) this.box.focus();
				}.bind(this)
			}
		}).inject(this.contentBox);
		if(!this.options.overlayAll) {
			this.overlay.setStyle('top',(this.title ? this.title.getSize().y - 1: 0));
		}
		
		//create initial buttons
		if(this.options.buttons.length) {
			this.options.buttons.each(function(button) {
				this.addButton(button.title,button.event,button.color);
			},this);
		}
		
		//focus node
		this.focusNode = this.box;
		
		return this;
	},

	// Manage parent and dependent instances of LightFace.
	attachDependent: function(oDependent) {
		if (instanceOf(oDependent, LightFace) && this.dependents.indexOf(oDependent) === -1) {
			this.dependents.push(oDependent);
			oDependent.attachParent(this);
		}
		return this;
	},
	attachParent: function(oParent) {
		if (instanceOf(oParent, LightFace)) {
			this.oParent = oParent;
			if (oParent.dependents.indexOf(this) === -1) {
				oParent.attachDependent(this);
			}
		}
		return this;
	},
	_closeDependents: function() {
		if (this.dependents.length) {
			this.dependents.each(function(oObj) {
				oObj.close(true, this.options.destroyOnClose);
			}, this);
//			this.dependents.empty();
		}
		return this;
	},
	_detachDependent: function(oDependent) {
		if (instanceOf(oDependent, LightFace)) {
			this.dependents.erase(oDependent);
			oDependent.oParent = null;
		}
		return this;
	},
	_detachParent: function() {
		if (this.oParent) {
			this.oParent._detachDependent(this);
			this.oParent = null;
		}
		return this;
	},

	// Manage buttons
	addButton: function(title,clickEvent,color) {
		this.footer.setStyle('display','block');
		var focusClass = 'lightfacefocus' + color;
		var label = new Element('label',{
			'class': color ? 'lightface' + color : '',
			events: {
				mousedown: function() {
					if(color) {
						label.addClass(focusClass);
						var ev = function() {
							label.removeClass(focusClass);
							document.id(document.body).removeEvent('mouseup',ev);
						};
						document.id(document.body).addEvent('mouseup',ev);
					}
				}
			}
		});
		this.buttons[title] = (new Element('input',{
			type: 'button',
			value: title,
			events: {
				click: (clickEvent || this.close).bind(this)
			}
		}).inject(label));
		label.inject(this.footer);
		return this;
	},
	showButton: function(title) {
		if(this.buttons[title]) this.buttons[title].removeClass('hiddenButton');
		return this.buttons[title];
	},
	hideButton: function(title) {
		if(this.buttons[title]) this.buttons[title].addClass('hiddenButton');
		return this.buttons[title];
	},
	
	// Open and close box
	close: function(fast, bDestroy) {
		if(this.isOpen) {
			this._closeDependents();
			if (fast) {
				this.box.setStyle('opacity', 0);
				this.box.setStyle('visibility', 'hidden');
				if (this.oMask) this.oMask.hide();
			}
			else {
				this.box.get('tween').start('opacity', 0).chain(function()
				{
					this.box.setStyle('visibility', 'hidden');
					if (this.oMask) this.oMask.hide();
					if (bDestroy || this.options.destroyOnClose) this.destroy();
				}.bind(this));
			}
			this.fireEvent('close');
			this._detachEvents();
			this.isOpen = false;
			if (fast && (bDestroy || this.options.destroyOnClose)) this.destroy();
		}
		return this;
	},
	
	open: function(fast) {
		if(!this.isOpen) {
			if (this.oMask) this.oMask.show();
			this.box.setStyle('visibility', 'visible');
			this.box[fast ? 'setStyle' : 'tween']('opacity', 1);
			if(this.resizeOnOpen) this._resize();
			this.fireEvent('open');
			this._attachEvents();
			(function() {
				this._setFocus();
			}).bind(this).delay(this.options.fadeDuration + 10);
			this.isOpen = true;
		}
		return this;
	},
	
	_setFocus: function() {
		this.focusNode.setAttribute('tabIndex',0);
		this.focusNode.focus();
	},
	
	// Show and hide overlay
	fade: function(fade,delay) {
		this._ie6Size();
		(function() {
			this.overlay.setStyle('visibility', 'visible');
			this.overlay.setStyle('opacity',fade || 1);
		}.bind(this)).delay(delay || 0);
		this.fireEvent('fade');
		return this;
	},
	unfade: function(delay) {
		(function() {
			this.overlay.get('tween').start('opacity', 0).chain(function(){ this.element.setStyle('visibility', 'hidden'); });
		}.bind(this)).delay(delay || this.options.fadeDelay);
		this.fireEvent('unfade');
		return this;
	},
	_ie6Size: function() {
		if(this.ie6) {
			var size = this.contentBox.getSize();
			var titleHeight = (this.options.overlayAll || !this.title) ? 0 : this.title.getSize().y;
			this.overlay.setStyles({
				height: size.y - titleHeight,
				width: size.x
			});
		}
	},
	
	// Loads content
	load: function(content,title) {
		if(content) this.messageBox.set('html',content);
		title = title || this.options.title;
		if(title) this.title.set('html', title).setStyle('display', 'block');
		else this.title.setStyle('display', 'none');
		this.fireEvent('complete');
		return this;
	},
	
	// Attaches events when opened
	_attachEvents: function() {
		this.keyEvent = function(e){
			if(this.options.keys[e.key]) this.options.keys[e.key].call(this);
		}.bind(this);
		this.focusNode.addEvent('keyup',this.keyEvent);
		
		this.resizeEvent = this.options.constrain ? function(e) { 
			this._resize(); 
		}.bind(this) : function() { 
			this._position(); 
		}.bind(this);
		window.addEvent('resize',this.resizeEvent);
		
		if(this.options.resetOnScroll) {
			this.scrollEvent = function() {
				this._position();
			}.bind(this);
			window.addEvent('scroll',this.scrollEvent);
		}
		
		return this;
	},
	
	// Detaches events upon close
	_detachEvents: function() {
		this.focusNode.removeEvent('keyup',this.keyEvent);
		window.removeEvent('resize',this.resizeEvent);
		if(this.scrollEvent) window.removeEvent('scroll',this.scrollEvent);
		return this;
	},
	
	// Repositions the box
	_position: function() {
		var windowSize = window.getSize(), 
			scrollSize = window.getScroll(), 
			boxSize = this.box.getSize();
		this.box.setStyles({
			left: scrollSize.x + ((windowSize.x - boxSize.x) / 2),
			top: scrollSize.y + ((windowSize.y - boxSize.y) / 2)
		});
		this._ie6Size();
		return this;
	},
	
	// Resizes the box, then positions it
	_resize: function() {
		var height = this.options.height;
		if(height == 'auto') {
			//get the height of the content box
			var max = window.getSize().y - this.options.pad;
			if(this.contentBox.getSize().y > max) height = max;
		}
		this.messageBox.setStyle('height',height);
		this._position();
	},
	
	// Expose message box
	toElement: function () {
		return this.messageBox;
	},
	
	// Expose entire modal box
	getBox: function() {
		return this.box;
	},
	
	// Cleanup
	destroy: function() {
		this._closeDependents();
		this._detachParent();
		this._detachEvents();
		if (this.oMask) this.oMask.destroy();
		Object.each(this.buttons, function(button) {
			button.removeEvents('click');
		});
		this.box.dispose();
		delete this.box;
	}
});